public/test/test-utils.tsx

```diff
@@ -0,0 +1,107 @@
+import { render, RenderOptions } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { KBarProvider } from 'kbar';
+import React, { ComponentProps, Fragment, PropsWithChildren } from 'react';
+import { Provider } from 'react-redux';
+import { MemoryRouter } from 'react-router-dom';
+import { PreloadedState } from 'redux';
+import { getGrafanaContextMock } from 'test/mocks/getGrafanaContextMock';
+
+import { config } from '@grafana/runtime';
+import { ErrorBoundaryAlert } from '@grafana/ui';
+import { GrafanaContext, GrafanaContextType } from 'app/core/context/GrafanaContext';
+import { ModalsContextProvider } from 'app/core/context/ModalsContextProvider';
+import { ThemeProvider } from 'app/core/utils/ConfigProvider';
+import { configureStore } from 'app/store/configureStore';
+import { StoreState } from 'app/types/store';
+
+interface ExtendedRenderOptions extends RenderOptions {
+  /**
+   * Optional store to use for rendering. If not provided, a fresh store will be generated
+   * via `configureStore` method
+   */
+  store?: ToolkitStore;
+  /**
+   * Partial state to use for preloading store when rendering tests
+   */
+  preloadedState?: PreloadedState<StoreState>;
+  /**
+   * Should the wrapper be generated with a wrapping Router component?
+   * Useful if you're testing something that needs more nuanced routing behaviour
+   * and you want full control over it instead
+   */
+  renderWithRouter?: boolean;
+  /**
+   * Props to pass to `MemoryRouter`, if being used
+   */
+  routerOptions?: ComponentProps<typeof MemoryRouter>;
+}
+
+/**
+ * Get a wrapper component that implements all of the providers that components
+ * within the app will need
+ */
+const getWrapper = ({
+  store,
+  renderWithRouter,
+  routerOptions,
+  grafanaContext,
+}: ExtendedRenderOptions & {
+  grafanaContext?: GrafanaContextType;
+}) => {
+  const reduxStore = store || configureStore();
+  /**
+   * Conditional router - either a MemoryRouter or just a Fragment
+   */
+  const PotentialRouter = renderWithRouter ? MemoryRouter : Fragment;
+
+  const context = {
+    ...getGrafanaContextMock(),
+    ...grafanaContext,
+  };
+
+  /**
+   * Returns a wrapper that should (closely) match the main `AppWrapper`, so any tests are rendering
+   * in mostly the same providers as a "real" hierarchy
+   */
+  return function Wrapper({ children }: PropsWithChildren) {
+    return (
+      <Provider store={reduxStore}>
+        <ErrorBoundaryAlert style="page">
+          <GrafanaContext.Provider value={context}>
+            <ThemeProvider value={config.theme2}>
+              <KBarProvider>
+                <PotentialRouter {...routerOptions}>
+                  <ModalsContextProvider>{children}</ModalsContextProvider>
+                </PotentialRouter>
+              </KBarProvider>
+            </ThemeProvider>
+          </GrafanaContext.Provider>
+        </ErrorBoundaryAlert>
+      </Provider>
```

Comment:
I'm... skeptical that we always want all of these for all tests... but don't feel super strongly about this. I also don't really have a better suggestion.

Maybe one that we can get rid of here for sure is KBarProvider, and just continue to manually add it in like the two tests that use it?

I also have a slight personal preference for a pattern more like exporting this `render`, and then a `renderWithRouter` function you could use instead, but that doesn't really scale or compose any further so I also don't feel strongly about that.

Reply:
I think it's a fair point about `KBarProvider`, as that one does seem to be very rarely required for tests. I guess an argument for _keeping_ it might be in case any other component defines some keyboard shortcuts, then you would know that your test will be running with the KBar logic as well.

I guess the only restriction with omitting it is that you can't easily get it in the same place in the providers hierarchy without redefining all of the other ones 
i.e., `KBar` is in between `Theme` and `Router` in `AppWrapper`, and if the order mattered significantly then you'd have to import everything in your test - but that's the edge case (and I really don't think the order will matter for that anyway).

Regardless: I'm happy to remove it if you'd like :+1: